Add channel arg for max send message size and add message size filter.

markdroth opened the pull request:
This is part of addressing #7818.  It depends on #7839.

markdroth commented on include/grpc/impl/codegen/grpc_types.h:
```diff
@@ -127,7 +127,11 @@ typedef struct {
 #define GRPC_ARG_MAX_CONCURRENT_STREAMS "grpc.max_concurrent_streams"
 /** Maximum message length that the channel can receive. Int valued, bytes. */
-#define GRPC_ARG_MAX_MESSAGE_LENGTH "grpc.max_message_length"
+#define GRPC_ARG_MAX_RECEIVE_MESSAGE_LENGTH "grpc.max_receive_message_length"
```
Question: Is it okay to change the value of this parameter?  I've added a `#define' to provide API backward-compatibility, but changing the string here affects ABI compatibility.  I'm not sure how big of an issue that is.

markdroth replied:
@ctiller, can you confirm that this is okay?  Thanks!

markdroth replied:
Just to close this loop, I chatted with Craig, and he said this should be fine.

ctiller commented on src/core/lib/channel/message_size_filter.c:
```diff
@@ -0,0 +1,177 @@
+  chand->max_send_size = DEFAULT_MAX_MESSAGE_LENGTH;
+  chand->max_recv_size = DEFAULT_MAX_MESSAGE_LENGTH;
+  for (size_t i = 0; i < args->channel_args->num_args; ++i) {
+    if (strcmp(args->channel_args->args[i].key,
```
It's really getting time for a utility or two here... something like:

```
typedef struct {
  int default_value;
  int min_value;
  int max_value;
} grpc_integer_options;
int grpc_channel_args_get_integer(grpc_channel_args *args, grpc_integer_options options);
```

markdroth replied:
Done.

markdroth commented:
Reminder: Please review #7839 before this one.  (In retrospect, I should probably not have merged those changes into this branch, since they're largely independent, but I didn't realize that at the time.)

Thanks!

ctiller commented on src/core/ext/transport/chttp2/transport/chttp2_transport.c:
```diff
@@ -337,76 +339,65 @@ static void init_transport(grpc_exec_ctx *exec_ctx, grpc_chttp2_transport *t,
-          gpr_log(GPR_ERROR, "%s: must be non-negative",
-                  GRPC_ARG_MAX_METADATA_SIZE);
-        } else {
+        const grpc_integer_options options = {-1, 0, INT_MAX};
```
This makes me insanely happy.

markdroth commented:
@apolcyn, could you please take a look at the objc test failures here?  I suspect that we may need to plumb the new max-send-size option through the objc API in order to make the tests work, but I don't speak objc, so I don't know how to do that myself.  If you could send me a PR (to the markdroth/grpc repo, so I can merge it into this PR), that would be great.  What do you think?

apolcyn commented:
@markdroth I've been working with c#, but not objc, unfortunately I'm not familiar with objc.

cc @jcanizales @makdharma do you know?

markdroth commented:
Sorry, Alex, looks like I had the wrong person.

Jorge, is this something you can help me with?  Thanks!

jcanizales commented:
The error is this:

```
error: -[InteropTestsLocalCleartext testResponsesOver4MBAreAcceptedIfOptedIn] : ((error) == nil) failed: "Error Domain=io.grpc Code=3 "Sent message larger than max (5242890 vs. 4194304)"
```

It means this change would break people that try to set the size limit to a value higher than the default.

markdroth commented:
Yeah, I can tell that.  What I suspect is happening is that since I've added a new parameter for the max send size, the test is failing on the sending side, because the limit was raised on the receiving side, but not the sending side.  So I think we need to plumb the new option through the objc API, and then change the tests to increase both limits.  What do you think?

jcanizales commented:
You mean the server is producing that error? This is sending requests to the C++ interop server, so we'd need to figure out how to let it send big responses.

markdroth commented:
Hmm, I had somehow missed the implications of the fact that this is an interop test.  If the objective-c test is just getting a response from the server, then I agree that the send limit would have to be raised on the server side.

That having been said, it's not clear to me why this limit is being checked in an interop test to begin with.  The interop test descriptions doc doesn't say anything about a test for max message size:

https://github.com/grpc/grpc/blob/master/doc/interop-test-descriptions.md

Furthermore, since no interop tests failed in any other languages, I assume that none of the other languages are doing this.  So why are we doing it for objective-c but none of the others?

jcanizales commented:
We don't have ObjC servers, so we use that one for non-unit testing. I haven't looked into what the other languages do to test this, but the tests are:

1. You can receive responses of 4MB no problem without doing anything special.
2. The error message for receiving responses over the limit is meaningful.
3. You can actually receive responses over 4MB by setting the limit to a higher value.

The only way to test the third one is by having a server allowed to send bigger responses.

markdroth commented:
It seems like a bad idea to test this kind of thing from the interop test framework, because that doesn't allow you to change parameters on the client and server on a per-test basis.  For example, I can presumably fix the immediate problem by increasing the send-size limit on the C++ server.  However, if we then also want to add a test that the server enforces its send-size limit, we can't change the server's send-size limit without breaking the objective-c tests.

For now, I'll go ahead and add a way to increase the send-size limit in the C++ API and change the interop tests to run with a higher limit.  However, in parallel with that, can you please move these tests out of the interop framework and into some other harness?  Thanks!

jcanizales commented:
> It seems like a bad idea to test this kind of thing from the interop test framework, because that doesn't allow you to change parameters on the client and server on a per-test basis.

The way the ObjC tests that require a server are run is a shell script starts the C++ interop server locally (actually two times, one with TLS and one without), and XCode is told to execute the tests in the iPhone simulator. There is no way to change parameters on a per-test basis anyway, as the tests are executed by the simulator, which has no way to start servers in the host machine. This is already separate from the interop framework we have for the other languages, in which clients and servers can be brought up in GCE.

> However, if we then also want to add a test that the server enforces its send-size limit, we can't change the server's send-size limit without breaking the objective-c tests.

A way to achieve this is adding a flag to the C++ server to set the limit. Then ObjC tests can run the server with a higher limit than the default.

markdroth commented:
Okay, thanks for the clarification.  I didn't realize that this was a non-interop test that happened to use the C++ interop server.  The setup makes more sense now.

I've added the flag to the C++ interop server and changed the objc test harness to use it.  Let's see if that fixes the problem.

jcanizales commented:
[That ObjC failure in Jenkins](https://grpc-testing.appspot.com/job/gRPC_pull_requests/config=dbg,language=objc,platform=macos/12664/testReport/junit/(root)/tests/objc_tests/) was [faced and solved by a stackoverflower](http://stackoverflow.com/questions/37195135/xcodebuild-mt-iphonesimulator-could-not-launch-simulator-600?rq=1) by restarting the Mac slave.

markdroth commented:
I suspect that the remaining test failures are flakes.

dgquintas commented on include/grpc++/server_builder.h:
```diff
@@ -78,12 +78,23 @@ class ServerBuilder {
     return *this;
   }
 
+  /// For backward compatibility.
```
We may want to add a `\deprecated` in here (viz. https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmddeprecated)

markdroth replied:
Done.

dgquintas commented on include/grpc/impl/codegen/grpc_types.h:
```diff
@@ -127,7 +127,11 @@ typedef struct {
 /** Maximum message length that the channel can receive. Int valued, bytes. */
-#define GRPC_ARG_MAX_MESSAGE_LENGTH "grpc.max_message_length"
+#define GRPC_ARG_MAX_RECEIVE_MESSAGE_LENGTH "grpc.max_receive_message_length"
+/** Backward compatibility. */
```
`\deprecated`?

markdroth replied:
Done.

dgquintas commented on src/core/lib/channel/channel_args.h:
```diff
@@ -89,4 +89,12 @@ uint32_t grpc_channel_args_compression_algorithm_get_states(
 int grpc_channel_args_compare(const grpc_channel_args *a,
                               const grpc_channel_args *b);
 
+/** Returns the value of \a arg, subject to the contraints in \a options. */
```
This comment seems misplaced: as is, it's documenting the `grpc_integer_options` enum instead of the `grpc_channel_arg_get_integer` function.

dgquintas replied:
Meaning, doxygen will generate the wrong docs.

markdroth replied:
Moved the comment down to just above the function declaration.

dgquintas commented on src/core/lib/channel/message_size_filter.c:
```diff
@@ -0,0 +1,168 @@
+      (*calld->recv_message)->length > chand->max_recv_size) {
+    char* message_string;
+    gpr_asprintf(&message_string,
+                 "Received message larger than max (%lu vs. %lu)",
```
why not `%zu` and save the cast?

markdroth replied:
Done.

dgquintas commented on src/core/lib/channel/message_size_filter.c:
```diff
@@ -0,0 +1,168 @@
+        exec_ctx, elem, GRPC_STATUS_INVALID_ARGUMENT, &message);
+  }
+  // Invoke the next callback.
+  calld->next_recv_message_ready->cb(
```
Why not use the canonical way of calling closures? Ie, 
```
grpc_exec_ctx_sched(exec_ctx, calld->next_recv_message_ready, error, NULL);
```
this is safer, in ways recently described by @ctiller

markdroth replied:
I think that when I originally wrote that, I was thinking that `calld->next_recv_message_ready` might go out of scope by the time the callback is executed.  But I think I was wrong about that, because looking at the code now, I don't think that's possible.

Done.

dgquintas commented on src/core/lib/channel/message_size_filter.c:
```diff
@@ -0,0 +1,168 @@
+  if (op->send_message != NULL &&
+      op->send_message->length > chand->max_send_size) {
+    char* message_string;
+    gpr_asprintf(&message_string, "Sent message larger than max (%lu vs. %lu)",
```
`s/lu/zu`?

markdroth replied:
Done.

dgquintas commented on src/cpp/server/server_builder.cc:
```diff
@@ -160,8 +162,11 @@ std::unique_ptr<Server> ServerBuilder::BuildAndStart() {
-    args.SetInt(GRPC_ARG_MAX_MESSAGE_LENGTH, max_message_size_);
+  if (max_receive_message_size_ > 0) {
+    args.SetInt(GRPC_ARG_MAX_RECEIVE_MESSAGE_LENGTH, max_receive_message_size_);
+  }
```
should we error/log if <= 0?

markdroth replied:
This error will get logged anyway deeper in the stack (inside of `grpc_channel_arg_get_integer()`).

dgquintas commented on test/core/end2end/tests/max_message_length.c:
```diff
@@ -98,19 +98,22 @@ static void end_test(grpc_end2end_test_fixture *f) {
-  grpc_channel_args server_args;
+  grpc_arg channel_arg;
+  grpc_channel_args channel_args;
   grpc_call *c;
```
why not initialize `c` to `NULL` as well while we are here

markdroth replied:
Done.

dgquintas commented on test/cpp/interop/interop_server.cc:
```diff
@@ -321,6 +322,9 @@ void grpc::testing::interop::RunServer(
   ServerBuilder builder;
   builder.RegisterService(&service);
   builder.AddListeningPort(server_address.str(), creds);
+  if (FLAGS_max_send_message_size >= 0) {
```
https://github.com/grpc/grpc/pull/7846/files#diff-5dd9c06965eebe9bf2d14a72f1c8d064R165 ignores values <= 0, but 0 is accepted here.

markdroth replied:
Fixed the other location to accept a value of 0.

markdroth commented on doc/core/pending_api_cleanups.md:
```diff
@@ -0,0 +1,15 @@
+There are times when we make changes that include a temporary shim for
```
@ctiller, here's the file you requested that I add yesterday.

@nicolasnoble, please keep this file in mind whenever the next release happens.

jcanizales replied:
Why bump the *minor* version number when making breaking changes, instead of the major number as semantic versioning requires? That messes with users of dependency managers that understand semantic versioning.

markdroth replied:
Whoops -- that was a brain-fart on my part.  Fixed.

markdroth commented:
It looks like the remaining failures here are unrelated to this PR, so I'll go ahead and merge this.
